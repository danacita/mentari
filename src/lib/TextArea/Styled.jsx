import styled, { css } from 'styled-components';

import Color from '../Styles/bases/Color';
import { Transition } from '../Styles/utils';

const TextAreaWrapper = styled.div`
  position: relative;
  font-size: 16px;
  textarea:-webkit-autofill {
    box-shadow: 0 0 0px 1000px white inset;
    -webkit-box-shadow: 0 0 0px 1000px white inset;
  }
  textarea {
    position: relative;
    border: 1px solid ${Color.grey191};
    line-height: 1.6;
    font-size: 16px;
    display: block;
    min-height: 46px;
    width: 100%;
    padding: 8px 16px;
    border-radius: 4px;
    ${Transition('all ease-in-out 0.2s')} outline: none;
    &::placeholder {
      color: white;
    }
    &:focus,
    &:active {
      border-color: ${Color.blue};
      padding-top: 14px;
      + label {
        top: -10px;
        color: ${Color.blue};
        font-size: 12px;
        font-weight: 600;
      }
      &::placeholder {
        color: ${Color.grey191};
      }
    }
    + label {
      position: absolute;
      z-index: 2;
      left: 8px;
      background-color: ${Color.white};
      padding: 0 8px;
      top: 12px;
      ${Transition('all ease-in-out 0.2s')};
      cursor: text;
      color: ${Color.grey191};
      height: 24px;
      display: flex;
      align-items: center;
    }
    &:disabled {
      background-color: ${Color.white};
      cursor: not-allowed;
      color: ${Color.grey216};
      border-color: ${Color.grey239};
      &::placeholder {
        color: ${Color.grey239};
        opacity: 0;
      }
      + label {
        background-color: ${Color.white};
        color: ${Color.grey216};
        cursor: not-allowed;
      }
      &:active {
        padding-top: 14px;
        border-color: ${Color.grey191};
        + label {
          top: 12px;
          color: ${Color.grey191};
          font-size: 14px;
          font-weight: 400;
        }
        &::placeholder {
          color: ${Color.grey191};
        }
      }
    }
  }
  ${props =>
    props.valLength > 0
      ? css`
          textarea {
            padding-top: 14px;
            + label {
              top: -10px;
              color: ${Color.grey191};
              font-size: 12px;
              font-weight: 600;
            }
          }
        `
      : null};
  ${props =>
    props.touched && props.error
      ? css`
          textarea {
            border: 1px solid ${Color.red};
            &:focus,
            &:hover,
            &:active {
              border-color: ${Color.red};
              + label {
                color: ${Color.red};
              }
            }
            + label {
              color: ${Color.red};
            }
          }
        `
      : null};
  ${props =>
    props.hasPrefix
      ? css`
          textarea {
            padding-left: 44px;
          }
        `
      : null};
`;

const TextAreaPrefix = styled.div`
  position: absolute;
  z-index: 2;
  top: 12px;
  left: 14px;
`;

export { TextAreaWrapper, TextAreaPrefix };
