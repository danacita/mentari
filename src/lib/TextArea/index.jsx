/* eslint-disable jsx-a11y/label-has-for */
import React from 'react';
import PropTypes from 'prop-types';

import { TextAreaWrapper, TextAreaPrefix } from './Styled';
import { Margin } from '../Styles/utils';
import Color from '../Styles/bases/Color';

const propTypes = {
  id: PropTypes.string,
  testid: PropTypes.string,
  name: PropTypes.string,
  label: PropTypes.string,
  placeholder: PropTypes.string,
  disabled: PropTypes.bool,
  prefix: PropTypes.node,
  // input prop is passed by redux-form to keep field in sync with state
  input: PropTypes.shape({
    value: PropTypes.string,
    onBlur: PropTypes.func,
    onChange: PropTypes.func
  }).isRequired,
  meta: PropTypes.shape({
    error: PropTypes.string,
    touched: PropTypes.bool
  }).isRequired
};

const defaultProps = {
  id: '',
  testid: '',
  name: '',
  placeholder: null,
  disabled: false,
  prefix: '',
  label: ''
};

const TextArea = ({
  id,
  name,
  testid,
  label,
  placeholder,
  prefix,
  disabled,
  input,
  meta: { error, touched }
}) => (
  <TextAreaWrapper
    valLength={input.value && input.value.length}
    error={error}
    touched={touched}
    hasPrefix={!!prefix}
    disabled={disabled}
  >
    {prefix ? <TextAreaPrefix>{prefix}</TextAreaPrefix> : null}
    <textarea
      id={id}
      data-testid={testid}
      name={name}
      rows="3"
      placeholder={placeholder}
      value={disabled ? '' : input.value === undefined ? '' : input.value}
      onBlur={input.onBlur}
      onChange={input.onChange}
      disabled={disabled}
    />
    <label htmlFor={id}>{label}</label>
    {touched && error && (
      <Margin top={8}>
        <span style={{ color: Color.red }}>{error}</span>
      </Margin>
    )}
  </TextAreaWrapper>
);

TextArea.propTypes = propTypes;
TextArea.defaultProps = defaultProps;

export default TextArea;
