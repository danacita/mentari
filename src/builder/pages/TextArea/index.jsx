import React, { Component } from 'react';
import SyntaxHighlighter from 'react-syntax-highlighter';
import { atomOneDarkReasonable } from 'react-syntax-highlighter/dist/esm/styles/hljs';
import { ContentWithRightBar } from '../Styled';
import { Margin, TextField, TextArea, Row, Col, Checkbox } from '../../../lib';
import PropsShared from '../../shared/Props';
import DemoShared from '../../shared/Demo';
import CodeShared from '../../shared/Code';
import RightbarShared from '../../shared/Rightbar';

class TextAreaPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      demoTextArea: '',
      handleIdTextArea: 'Id',
      handleNameTextArea: 'Name',
      handleLabelTextArea: 'Label',
      handlePlaceholderTextArea: 'Placeholder',
      handleDisableTextArea: false
    };
  }

  render() {
    const {
      demoTextArea,
      handleIdTextArea,
      handleNameTextArea,
      handleLabelTextArea,
      handlePlaceholderTextArea,
      handleDisableTextArea
    } = this.state;
    const codeString = `import { TextArea } from '@danacita/mentari';\n<TextArea id="${handleIdTextArea}" name="${handleNameTextArea}" label="${handleLabelTextArea}" placeholder="${handlePlaceholderTextArea}" ${
      handleDisableTextArea ? 'disabled ' : ''
    }/>`;
    return (
      <ContentWithRightBar>
        <h1>Textfield</h1>
        <p>Lorem ipsum dolor, sit amet consectetur adipisicing elit.</p>
        <Margin top={24}>
          <PropsShared>
            <li>
              <span>id</span>
              <span>Apply a id to the control element</span>
            </li>
            <li>
              <span>name</span>
              <span>Specifies a name for a Textarea</span>
            </li>
            <li>
              <span>label</span>
              <span>Defines label name Textarea</span>
            </li>
            <li>
              <span>placeholder</span>
              <span>Apply a short hint that describes the expected value when input focus</span>
            </li>
            <li>
              <span>disabled</span>
              <span>Apply disabled state of Textarea for unusable and unclickable</span>
            </li>
          </PropsShared>
        </Margin>
        <DemoShared>
          <h4>Demo</h4>
          <p>You can change the props via the component builder on your right</p>
          <Row>
            <Col sm={12} lg={4}>
              <Margin top={16}>
                <TextArea
                  id={handleIdTextArea}
                  name={handleNameTextArea}
                  label={handleLabelTextArea}
                  placeholder={handlePlaceholderTextArea}
                  disabled={handleDisableTextArea}
                  withRedux={false}
                  value={demoTextArea}
                  onChange={e => this.setState({ demoTextArea: e.target.value })}
                />
              </Margin>
            </Col>
          </Row>
        </DemoShared>
        <h4>Code</h4>
        <Margin top={16}>
          <CodeShared copyCode={codeString}>
            <SyntaxHighlighter language="jsx" style={atomOneDarkReasonable} showLineNumbers>
              {codeString}
            </SyntaxHighlighter>
          </CodeShared>
        </Margin>
        <RightbarShared title="Textarea">
          <Margin bottom={24}>
            <TextField
              autoFocus
              id="handleIdTextArea"
              label="Textarea Id"
              withRedux={false}
              value={handleIdTextArea}
              onChange={e => this.setState({ handleIdTextArea: e.target.value })}
            />
          </Margin>
          <Margin bottom={24}>
            <TextField
              id="handleNameTextArea"
              label="Textarea Name"
              withRedux={false}
              value={handleNameTextArea}
              onChange={e => this.setState({ handleNameTextArea: e.target.value })}
            />
          </Margin>
          <Margin bottom={24}>
            <TextField
              id="handleLabelTextArea"
              label="Textarea Label"
              withRedux={false}
              value={handleLabelTextArea}
              onChange={e => this.setState({ handleLabelTextArea: e.target.value })}
            />
          </Margin>
          <Margin bottom={24}>
            <TextField
              id="handlePlaceholderTextArea"
              label="Textarea Placeholder"
              withRedux={false}
              value={handlePlaceholderTextArea}
              onChange={e => this.setState({ handlePlaceholderTextArea: e.target.value })}
            />
          </Margin>
          <Margin bottom={16}>
            <h4>State of Textarea</h4>
          </Margin>
          <Margin bottom={24}>
            <Checkbox
              id="disable"
              name="disable"
              label="Disable"
              checked={handleDisableTextArea}
              withRedux={false}
              onChange={e => this.setState({ handleDisableTextArea: e.target.checked })}
            />
          </Margin>
        </RightbarShared>
      </ContentWithRightBar>
    );
  }
}

export default TextAreaPage;
