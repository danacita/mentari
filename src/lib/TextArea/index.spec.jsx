import React from 'react';
import renderer from 'react-test-renderer';

import TextArea from '.';

describe('TextArea', () => {
  it('renders correctly', () => {
    const input = {
      value: 'Value',
      onBlur: () => {},
      onChange: () => {}
    };

    const meta = {
      error: '',
      touched: false
    };

    const tree = renderer.create(<TextArea input={input} meta={meta} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
