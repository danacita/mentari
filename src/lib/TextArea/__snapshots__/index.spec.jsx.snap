// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea renders correctly 1`] = `
<div
  className="sc-htpNat ivXWwS"
  disabled={false}
>
  <textarea
    data-testid=""
    disabled={false}
    id=""
    name=""
    onBlur={[Function]}
    onChange={[Function]}
    placeholder={null}
    rows="3"
    value="Value"
  />
  <label
    htmlFor=""
  >
    
  </label>
</div>
`;
